packages/ra-ui-materialui/src/list/datagrid/DatagridConfigurable.tsx

```diff
@@ -0,0 +1,136 @@
+import * as React from 'react';
+import { useResourceContext, usePreference, useStore } from 'ra-core';
+
+import { Configurable } from '../../preferences';
+import { Datagrid, DatagridProps } from './Datagrid';
+import { DatagridEditor } from './DatagridEditor';
+
+/**
+ * A Datagrid that users can customize in configuration mode
+ *
+ * @example
+ * import {
+ *     List,
+ *     DatagridConfigurable,
+ *     TextField,
+ * } from 'react-admin';
+ *
+ * export const PostList = () => (
+ *     <List>
+ *         <DatagridConfigurable>
+ *             <TextField source="id" />
+ *             <TextField source="title" />
+ *             <TextField source="author" />
+ *             <TextField source="year" />
+ *         </DatagridConfigurable>
+ *     </List>
+ * );
+ */
+export const DatagridConfigurable = ({
+    preferenceKey,
+    omit,
+    ...props
+}: DatagridConfigurableProps) => {
+    if (props.optimized) {
+        throw new Error(
+            'DatagridConfigurable does not support the optimized prop'
+        );
+    }
+    const resource = useResourceContext(props);
+    const finalPreferenceKey = preferenceKey || `${resource}.datagrid`;
+
+    const [availableColumns, setAvailableColumns] = useStore<
+        { index: string; source: string; label?: string }[]
+    >(`preferences.${finalPreferenceKey}.availableColumns`, []);
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [_, setOmit] = useStore<string[]>(
+        `preferences.${finalPreferenceKey}.omit`,
+        omit
+    );
+
+    React.useEffect(() => {
+        // first render, or the preference have been cleared
+        const columns = React.Children.map(props.children, (child, index) =>
+            React.isValidElement(child) &&
+            (child.props.source || child.props.label)
+                ? {
+                      index: String(index),
+                      source: child.props.source,
+                      label: child.props.label,
+                  }
+                : null
+        ).filter(column => column != null);
+        if (JSON.stringify(columns) !== JSON.stringify(availableColumns)) {
+            setAvailableColumns(columns);
+            setOmit(omit);
+        }
+    }, [availableColumns]); // eslint-disable-line react-hooks/exhaustive-deps
+
+    return (
+        <Configurable
+            editor={<DatagridEditor>{props.children}</DatagridEditor>}
+            preferenceKey={finalPreferenceKey}
+            sx={{
+                display: 'block',
+                '& .MuiBadge-root': { display: 'flex' },
+                '& .RaDatagrid-root': { flex: 1 },
+                '& .MuiBadge-badge': { zIndex: 2 },
```

Comment:
```suggestion
                '& .MuiBadge-badge': { zIndex: 2 },
                minHeight: 2,
```

> @WiXSL Right. I'm not sure how to solve this, though. Any suggestion?

I tried this and it seems to work (doesn't work with `minHeight: 1`):

![screen](https://user-images.githubusercontent.com/8268610/196950653-5d2f381c-5513-41a8-b783-0b5b0c5fc898.png)

Reply:
Brilliant!